File: frontend/src/lib/api/schemas/prompts.ts
import { z } from "zod";
import { PromptRunStatus } from "@/types/enums";

const promptRunStatusSchema = z.enum([PromptRunStatus.SUCCEEDED, PromptRunStatus.FAILED]);

export const promptTemplateReadSchema = z.object({
  id: z.string().uuid(),
  agent: z.string(),
  name: z.string(),
});
export type PromptTemplateRead = z.infer<typeof promptTemplateReadSchema>;

export const promptVersionReadSchema = z.object({
  id: z.string().uuid(),
  template_id: z.string().uuid(),
  version: z.number(),
  content: z.string(),
  front_matter: z.record(z.string(), z.unknown()),
  model_hints: z.record(z.string(), z.unknown()),
  variables: z.record(z.string(), z.unknown()),
});
export type PromptVersionRead = z.infer<typeof promptVersionReadSchema>;

export const promptRunReadSchema = z.object({
  id: z.string().uuid(),
  prompt_version_id: z.string().uuid(),
  project_id: z.string().uuid().nullable(),
  input_variables: z.record(z.string(), z.unknown()),
  rendered_prompt: z.string(),
  model_used: z.string(),
  status: promptRunStatusSchema,
  response: z.string().nullable(),
  latency_ms: z.number().nullable(),
  cost_usd: z.string().nullable(),
});
export type PromptRunRead = z.infer<typeof promptRunReadSchema>;

export type PromptVersionCreateRequest = {
  agent: string;
  name: string;
  content: string;
  front_matter?: Record<string, unknown>;
  model_hints?: Record<string, unknown>;
  variables?: Record<string, unknown>;
};

export type PromptRunCreateRequest = {
  prompt_version_id: string;
  input_variables: Record<string, unknown>;
  rendered_prompt: string;
  model_used: string;
  status: PromptRunStatus;
  project_id?: string | null;
  response?: string | null;
  latency_ms?: number | null;
  cost_usd?: string | null;
};

File: backend/src/ytforge/interfaces/api/schemas/prompts.py
from __future__ import annotations

import uuid
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import PromptRunStatus


class PromptTemplateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    agent: str
    name: str


class PromptVersionCreateRequest(BaseModel):
    agent: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1)
    front_matter: dict[str, Any] = Field(default_factory=dict)
    model_hints: dict[str, Any] = Field(default_factory=dict)
    variables: dict[str, Any] = Field(default_factory=dict)


class PromptVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    template_id: uuid.UUID
    version: int
    content: str
    front_matter: dict[str, Any]
    model_hints: dict[str, Any]
    variables: dict[str, Any]


class PromptRunCreateRequest(BaseModel):
    prompt_version_id: uuid.UUID
    input_variables: dict[str, Any]
    rendered_prompt: str
    model_used: str
    status: PromptRunStatus
    project_id: uuid.UUID | None = None
    response: str | None = None
    latency_ms: int | None = None
    cost_usd: Decimal | None = None


class PromptRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    prompt_version_id: uuid.UUID
    project_id: uuid.UUID | None
    input_variables: dict[str, Any]
    rendered_prompt: str
    model_used: str
    status: PromptRunStatus
    response: str | None
    latency_ms: int | None
    cost_usd: Decimal | None
